File: src/year2020/day1/day1.rs
//! [Day 1: Report Repair](https://adventofcode.com/2020/day/1)

struct Puzzle {
    expenses: Vec<u64>,
}

impl Puzzle {
    fn new(data: &str) -> Self {
        Self {
            expenses: data.lines().map(|s| s.parse().unwrap()).collect(),
        }
    }

    /// Solve part one.
    fn part1(&self) -> u64 {
        for i in &self.expenses {
            for j in &self.expenses {
                if i + j == 2020 {
                    return i * j;
                }
            }
        }
        0
    }

    /// Solve part two.
    fn part2(&self) -> u64 {
        for i in &self.expenses {
            for j in &self.expenses {
                for k in &self.expenses {
                    if i + j + k == 2020 {
                        return i * j * k;
                    }
                }
            }
        }
        0
    }
}

/// # Panics
#[must_use]
pub fn solve(data: &str) -> (u64, u64) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const TEST_INPUT: &str = include_str!("test.txt");

    #[test]
    fn test01() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part1(), 514579);
    }

    #[test]
    fn test02() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part2(), 241861950);
    }
}

File: src/year2020/day1/day1.py
#!/usr/bin/env python3
# [Day 1: Report Repair](https://adventofcode.com/2020/day/1)

import atexit
import sys
import time
from pathlib import Path

filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text().strip()
if "--elapsed" in sys.argv:
    sys.argv.remove("--elapsed")
    start_time_ns = time.time_ns()
    atexit.register(lambda: print(f"elapsed: {(time.time_ns() - start_time_ns) / 1_000_000}ms"))


lines = list(map(int, data.splitlines()))
length = len(lines)


def part1():
    for i in range(length):
        for j in range(length):
            if i != j:
                if lines[i] + lines[j] == 2020:
                    print(lines[i] * lines[j])
                    return


def part2():
    for i in range(length):
        for j in range(length):
            if i == j:
                continue
            p = lines[i] + lines[j]
            if p > 2020:
                continue
            for k in range(length):
                if i != j and i != k and j != k:
                    if p + lines[k] == 2020:
                        print(lines[i] * lines[j] * lines[k])
                        return


part1()
part2()
